### PR TITLE
Enhancement: Use friendsofphp/php-cs-fixer instead of StyleCI

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] requires `friendsof/php-cs-fixer` and removes the integration with StyleCI

### StyleCI

The integration with StyleCI has the following advantages:

- It applies coding standard issues automatically after a merge to `main`.

The integration with StyleCI has the following disadvantages:

- ~~A developer can not run StyleCI locally.~~ A developer can run StyleCI locally only by installing [a separate package](https://github.com/StyleCI/CLI), which however, requires PHP 7.2. We currently support PHP 7.1.
- If a developer wants to find out about the changes that would need to be applied, they have to navigate to a separate website. With the deep integration provided by GitHub Actions, that is not necessary anymore.
- StyleCI uses `friendsofphp/php-cs-fixer` under the hood, but adds obscurity on top. For example, it "invents" new fixers that can be enabled or enabled, but are in reality different configurations for existing fixers). Additionally, fixers that are available today in `friendsofphp/php-cs-fixer` cannot be configured in StyleCI without delay (if at all). Furthermore, we do not know which version of `friendsofphp/php-cs-fixer` is used under the hood.
- The integration with StyleCI violates the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment): When someone opens a pull request where the changes do not conform to the coding standard we configured, a reviewer can (and will be) surprised to see that the build passes. A reviewer might similarly (and rightfully) point out coding standard issues, which adds unnecessary noise to the conversation in a pull request.

### `friendsofphp/php-cs-fixer`

Requiring `friendsofphp/php-cs-fixer` as a dependency (similar to `phpstan/phpstan` and `vimeo/psalm`) alleviates all of these problems. We can use and configure any fixers that is available any time we want.

If we wanted to automatically apply coding standard issues (similar to StyleCI), we could easily do that with GitHub Actions. However, I would suggest to refrain from it. 

In my opinion, a pull request needs to be in mergeable state before merging it into the mainline of development, and this includes coding standard issues.

### CODE REVIEW

**Comment by pimjansen** [comment]
I agree with this to not be dependant on external. Running it locally is also a big advantage

**Comment by GrahamCampbell** [comment]
👎 here. We are just adding more maintenance for us here. We should stop messing with the code style, and just leave StyleCI to enforce it. Contributors and maintainers alike should not ever be running code style tooling.

**Comment by GrahamCampbell** [comment]
> A developer can not run StyleCI locally.

This is an advantage. Nobody should be wasting their time with that.

**Comment by GrahamCampbell** [comment]
Though, StyleCI can be run locally, if really wanted (https://github.com/StyleCI/CLI).

**Comment by localheinz** [comment]
@GrahamCampbell 

> This is an advantage. Nobody should be wasting their time with that.

I prefer to use and run `friendsofphp/php-cs-fixer` directly instead of integrating an external service or installing some tool that runs `friendsofphp/php-cs-fixer` for me.

But, I understand your opinion and your argumentation - after all, StyleCI is your pet project.

File: vendor-bin/php-cs-fixer/composer.json
Comment (pimjansen) [inline vendor-bin/php-cs-fixer/composer.json:4]: Require-dev?

Comment (GrahamCampbell) [reply vendor-bin/php-cs-fixer/composer.json:4]: no

**Comment by GrahamCampbell** [comment]
// cc @IonBazan, @mfn, @bramceulemans, @Nyholm

This change is a :-1: from me.

**Comment by bramceulemans** [comment]
I agree with it being more transparant to run php-cs-fixer. The idea of relying on an external service always worries me a bit personally.

I don't agree with the fact that developers shouldn't be doing code style. Using editorconfig file, phpstan and php-cs-fixer/StyleCI should be part of the IDE workflow.

File: vendor-bin/php-cs-fixer/composer.json
Comment (pimjansen) [reply vendor-bin/php-cs-fixer/composer.json:4]: Ah its vendor-bin nvm

**Comment by icanhazstring** [comment]
I am also in favor for php-cs-fixer as direct a dev dependency. The PR should be blocked unless everything is green.

Changing any code afterwards seems weird to me. I on my part only allow valid code to be merged into the main branch of repositories. So fine my me 👍

**Review by icanhazstring — APPROVED** [review]

**Review by Nyholm — COMMENTED** [review]
The code looks fine. 

I don't really mind either way. PHP-CS-fixer may have some benefits, but the same is true with StyleCI. If I were to decide, I would have solved this in a third way... 

However, I feel like I am a better resource if I focus on code changes.

**Review by bramceulemans — APPROVED** [review]

**Review by tomasnorre — APPROVED** [review]
I would always prefer a Style Checker locally. I want to provide as clean as possible code when pushing to what ever repository. 

I think it's important to see the changes done by a code style checker to learn from it. A lot of people just ignores code style, and if the CI does it for you, you will more and more often ignore what code you actually write.

My preferred option is having the phpcs for locally and what ever tool for pipeline taking care of the fixes not already done, but the two tools should follow the exact same ruleset of course.

**Comment by chris-doehring** [comment]
I see the benefit's of both strategies, but I also prefer using a "native" local tool for the following reasons:
* better learning experience for the contributers
* less noise within the PR for the reason that a reviewer might forget about the automatical adjustments after merging

**Comment by localheinz** [comment]
@GrahamCampbell 

Quick heads-up, after merging this pull request, StyleCI applied the following changes:

https://github.com/FakerPHP/Faker/commit/a57232a228d40ed9ec81599dbc1b22d868f0044b

Looks like a failure to me, to apply fixes when a configuration has been removed, and a great example for a violation of the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) as well.

**Comment by GrahamCampbell** [comment]
This is because StyleCI is still enabled on the repo, and uses the default configuration when none is provided. This is what people expect to happen, based on user feedback. The majority of StyleCI users never set any config.

**Comment by localheinz** [comment]
@GrahamCampbell 

> This is because StyleCI is still enabled on the repo, and uses the default configuration when none is provided. This is what people expect to happen, based on user feedback. The majority of StyleCI users never set any config.

Can you please disable the StyleCI integration that you introduced?

We do not want to use it anymore.

**Comment by driesvints** [comment]
I personally would have kept StyleCI for the same reasons @GrahamCampbell mentioned (regardless of biases) but I don't actively maintain this library so you should do whatever fits you best 🙂

File: .github/workflows/coding-standards.yml
Comment (sanmai) [inline .github/workflows/coding-standards.yml:19]: Should this be something like 7.4?

I'm late to the party, true.

File: .github/workflows/coding-standards.yml
Comment (sanmai) [inline .github/workflows/coding-standards.yml:29]: Is `intl` required here?

File: .github/workflows/coding-standards.yml
Comment (sanmai) [inline .github/workflows/coding-standards.yml:55]: `composer.lock` will be all new after each composer update, so this cache might not work as expected.

File: vendor-bin/php-cs-fixer/composer.json
Comment (sanmai) [inline vendor-bin/php-cs-fixer/composer.json:8]: It this unavoidable?

File: .github/workflows/coding-standards.yml
Comment (sanmai) [reply .github/workflows/coding-standards.yml:55]: And if we consider `vendor-bin` this should be `hashFiles('**/composer.*')`

**Pull request merged**
